### ISSUE TITLE
[Bug] Temporary files should be created at temporary dir

### ISSUE THREAD

**Issue by GitMensch** [OpenCobolIDE/OpenCobolIDE#168]
Reasons:
* maybe not possible to create them at the current file path (as seen below)
* if access is granted and something goes wrong you have temporary files that likely stay in a non temporary folder

```log
ERROR::pyqode.backend::something went bad with worker <function lint at 0x034C10C0>(data={'path': 'C:\\TEST.COB', 'code': [removed from log], 'encoding': 'cp1252'})
ERROR::pyqode.backend::Traceback (most recent call last):
ERROR::pyqode.backend::  File "c:\dev\pyqode.core\pyqode\core\backend\server.py", line 135, in _handle
ERROR::pyqode.backend::  File "C:\Dev\OpenCobolIDE\open_cobol_ide\linter.py", line 33, in lint
ERROR::pyqode.backend::PermissionError: [Errno 13] Permission denied: 'C:\\.oci1433411286.cbl'
```

This is what cobc/libcob does (pseudocode):

```c
if !(tmpdir = getenv ("TMPDIR")) {
#ifdef	_WIN32
   if (tmpdir = !getenv ("TEMP"))
      if !(tmpdir = getenv ("TMP"))
         if !(tmpdir = getenv ("USERPROFILE"))
            tmpdir = .;
#else
   if (tmpdir = !getenv ("TMP"))
      if !(tmpdir = getenv ("TEMP"))
         tmpdir = /tmp;
#endif

   setenv("TMPDIR", tmpdir);
}
```

For supporting relative paths to copybooks it should be enough to pass an additional -I $fullpath_of_current_file to the compiler

## System information

- OpenCobolIDE: 4.6.2
- Python: 3.4.3 (32bits)
- Qt: 5.4.1
- PyQt: 5.4.1
- pyqode.core: 2.7.0.dev2
- pyqode.cobol: 2.6.1
- pyqode.qt: 2.6.0
- pygments: 2.0.2

**Comment by ColinDuquesnoy** [comment]
This is what I was doing before #119. 

> This is what cobc/libcob does (pseudocode):

Python comes with battery included:

```python
import tempfile
tempfile.gettempdir()
```

> -I $fullpath_of_current_file

This is what I should have done in the first place! ;)

**Comment by ColinDuquesnoy** [comment]
The linter now does the following:

```
DEBUG::pyqode.backend::linter command: C:\Program Files (x86)\OpenCobolIDE\GnuCOBOL\bin\cobc.exe -fsyntax-only -IC:\Dev\OpenCobolIDE\test\testfiles -std=default -debug ocide_linter_1433595560.cbl
DEBUG::pyqode.backend::working directory: C:\Users\Colin\AppData\Local\Temp
DEBUG::pyqode.backend::linter raw output: ocide_linter_1433595560.cbl:8: Error: syntax error, unexpected ENVIRONMENT, expecting COMMON or RECURSIVE or INITIAL
DEBUG::pyqode.backend::linter parsed output: [('syntax error, unexpected ENVIRONMENT, expecting COMMON or RECURSIVE or INITIAL', 2, 7, 0, None, None, 'ocide_linter_1433595560.cbl')]
```

**Comment by GitMensch** [comment]
Looks much better.

Why not using the original filename (or to be perfect: use the original filename if the file is not in %temp%, the other elsewise).

It's very important for me to keep the original filename (if possible with original extension).

**Comment by ColinDuquesnoy** [comment]
> Why not using the original filename 

I use a hardcoded filename with a time stamp to avoid any conflict if multiple instance of OpenCobolIDE are running or when the user is working with split tabs and quicly change between tabs. But I could use ```%original_file_name_%timestamp.%original_file_extension``` if you prefer.

**Comment by GitMensch** [comment]
Let's switch to ```%original_file_name.%timestamp.%original_file_extension%``` ok?

**Comment by ColinDuquesnoy** [comment]
Ok, fine.

**Comment by ColinDuquesnoy** [comment]
Should be ok now:
```
DEBUG::pyqode.backend::linter command: /usr/local/bin/cobc -fsyntax-only -I/Users/Colin/Documents/OpenCobolIDE/test/testfiles -std=default -debug HelloWorld.1433611263.cbl
DEBUG::pyqode.backend::working directory: /var/folders/py/p42mn90d1v38bznwghsw49fr0000gn/T
DEBUG::pyqode.backend::linter raw output: HelloWorld.1433611263.cbl:8: Error: syntax error, unexpected IDENTIFICATION, expecting "FUNCTION-ID" or "PROGRAM-ID"
DEBUG::pyqode.backend::linter parsed output: [('syntax error, unexpected IDENTIFICATION, expecting "FUNCTION-ID" or "PROGRAM-ID"', 2, 7, 0, None, None, 'HelloWorld.1433611263.cbl')]
```

**Comment by GitMensch** [comment]
Looks good (I guess you've tested HelloWorld.COB, too). Thank you.

**Comment by ColinDuquesnoy** [comment]
;)

```
DEBUG::pyqode.backend::linter command: /usr/local/bin/cobc -fsyntax-only -I/Users/Colin -std=default -debug test.1433611710.cob
DEBUG::pyqode.backend::working directory: /var/folders/py/p42mn90d1v38bznwghsw49fr0000gn/T
DEBUG::pyqode.backend::linter raw output: test.1433611710.cob:7: Error: syntax error, unexpected IDENTIFICATION, expecting "FUNCTION-ID" or "PROGRAM-ID"
DEBUG::pyqode.backend::linter parsed output: [('syntax error, unexpected IDENTIFICATION, expecting "FUNCTION-ID" or "PROGRAM-ID"', 2, 6, 0, None, None, 'test.1433611710.cob')]
```

**Comment by GitMensch** [comment]
I really meant "COB" (original extension in upper case), but I guess this woks fine.

**Comment by ColinDuquesnoy** [comment]
Don't worry, it works:

```DEBUG::pyqode.backend::linter command: /usr/local/bin/cobc -fsyntax-only -I/Users/Colin -std=default -debug test.1433662951.COB```

**Comment by GitMensch** [comment]
The last change broke relative paths for copybooks via extra command line options :-(

-I ../cpy doesn't work any more

The only options where this is important for compile time checks are -conf (not that likely to be relative, but possible) and -I.

Both would have to be prefixed with $current_file_path + $path_sep to work if they are relative. This even applies to setting of COB_CONFIG_DIR and COB_COPY_DIR (but these are very unlikely to be relative).

**Comment by GitMensch** [comment]
Please reconsider changing the file name to not include the timestamp but "ocide" instead, as the number of temporary files currently increases a lot (they seem to not be deleted).

**Comment by ColinDuquesnoy** [comment]
You said:

> For supporting relative paths to copybooks it should be enough to pass an additional -I $fullpath_of_current_file to the compiler

This is what I've done:

```
/usr/bin/cobc -fsyntax-only -I/home/colin/Documents/OpenCobolIDE/test/testfiles -std=default -debug -I../tmp HelloWorld.1434362852.cbl
```

Is this wrong?

> Please reconsider changing the file name to not include the timestamp but "ocide" instead, as the number of temporary files currently increases a lot (they seem to not be deleted).

I am not going to remove the time stamp. I already said that but I will repeat: the time stamp is used to avoid conflicts when the user is working with multiple instances of the IDE or when he is using split tabs (and quickly switch from one editor to the other).

> they seem to not be deleted

They should, at least on my system they get deleted. https://github.com/OpenCobolIDE/OpenCobolIDE/blob/master/open_cobol_ide/linter.py#L78

**Comment by GitMensch** [comment]
> This is what I've done:
> [...]
> Is this wrong?

As written in #119 it's important to keep it in.

> I am not going to remove the time stamp. I already said that but I will repeat: the time stamp is used to avoid conflicts when the user is working with multiple instances of the IDE or when he is using split tabs (and quickly switch from one editor to the other).

Fine, as long as we don't pollute the file system

>> they seem to not be deleted

>They should, at least on my system they get deleted. https://github.com/OpenCobolIDE/OpenCobolIDE/blob/master/open_cobol_ide/linter.py#L78

Hm, maybe the thread is aborting before the line is executed?

**Comment by ColinDuquesnoy** [comment]
> Hm, maybe the thread is aborting before the line is executed?

This is not a thread but a background process that have the same life span as the open editor. If that process would end up before removing the file you would also loose code completion, navigation panels,.... There must be something else... I'll see if I can reproduce it on windows.

**Comment by ColinDuquesnoy** [comment]
Anything in the log?

**Comment by GitMensch** [comment]
Retested it and as always: if you want to reproduce something special it doesn't occur any more...
I suggest to close this and I'll bump it with a way to reproduce and/or more information as soon as I recognize this again.

**Comment by ColinDuquesnoy** [comment]
> I suggest to close this and I'll bump it with a way to reproduce and/or more information as soon as I recognize this again.

Ok.

**Comment by GitMensch** [comment]
The label invalid isn't correct as the original label "Bug" (closed) is correct ;-)

**Comment by ColinDuquesnoy** [comment]
Hmm at the moment it is invalid (should not be included in the next changelog since there is no fix). I've removed the bug label.

**Comment by GitMensch** [comment]
The "original" bug which is solved in 4.6.2 is "temporary files are created in workdir", isn't it?
So bug + closed + 4.6.2 looks best.

**Comment by ColinDuquesnoy** [comment]
> The "original" bug which is solved in 4.6.2 is "temporary files are created in workdir", isn't it?

Ah yes you're right ^^ Probably better to open a new issue if you stumble upon this bug again.
